### ISSUE TITLE
oneOf & not required

### ISSUE THREAD

**Issue by rjmk** [jquense/yup#14]
I was surprised to find out that `mixed.oneOf` doesn't allow `undefined`. I can imagine this being intended behaviour, but if not, perhaps it would be easiest to check if a given value is 'required' near the beginning of `_validate` and just exit positively if not?

Cheers

**Comment by jquense** [comment]
hey there thanks for the issue...I'm not sure I understand what your asking here. There is already a required() check that is more robust than just checking if its `undefined`. And I can't actually see any reason why you can't use `undefined` as a white or blacklisted value, it should work just fine. As with all the validations tho a positive result from one is not going to cancel the rest, that would be confusing..

In terms of validation order, yup is built with extensibility and flexibility of validation in mind. To that end we try in implement as much as possible using the same interface that a user has (i.e. the `.test()` method). The white and blacklists firing early like that is an artifact of an earlier design choice that I just haven't had the time or will to reimplement. That mixed with the fact that validations are async, a waterfall would actually generally be slower than the current method of running them in parallel.

**Comment by rjmk** [comment]
Hi, sorry, I guess I wasn't very clear.

If I have a schema like

```js
var mySchema = yup.object.shape({
  aThing: yup.string(),
  anotherThing: yup.oneOf(['a', 'b'])
})
```

And try it on the empty object, there'll be a validation error from `anotherThing`, whereas I would expect it to only worry that *if* there is a value, that it be one of the possible options.

**Comment by jquense** [comment]
Ok that makes more sense.  I don't think your expectation is wonky but it's also not unintentional either... this sort of lib is a huge bucket of tradeoffs and choices, since the types of interactions and expectations vary widely and aren't correct or incorrect. 

in this particular case tho I tend to think that empty values should be ignored here, like they are for other types of validations (like min or max). That being said there may be a good reason it's the way it is that I am forgetting.

**Comment by jquense** [comment]
fixed in 0.13
